caput/mpiarray.py

```
@@ -1036,6 +1036,177 @@ def _partition_io(self, skip=False, threshold=1.99):
+                dist_axis = len(results[0].shape) - 1
+
+        # Wrapping the results back into MPIArrays, distributed across the appropriate axis
+        ret = []
+
+        for result, output in zip(results, outputs):
+            if (
+                output is not None
+            ):  # case: results were placed in the array specified by `out`; return as is
+                ret.append(output)
+            else:
+                if (
+                    result.shape
+                ):  # case: the result is an array; convert back it into an MPIArray
+                    ret.append(MPIArray.wrap(result, axis=dist_axis))
+                else:  # case: result is a scalar; convert to 1-d vector, distributed across axis 0
+                    ret.append(MPIArray.wrap(np.reshape(result, (1, 1)), axis=0))
+
+        return ret[0] if len(ret) == 1 else tuple(ret)
+
+    # pylint: enable=inconsistent-return-statements
+
+    def __array_finalize__(self, obj):
+        """
+        Finalizes the creation of the MPIArray, when viewed.
+
+        In NumPy, ndarrays only go through the `__new__` when being instantiated.
+        For views and broadcast, they go through __array_finalize__.
+        https://numpy.org/doc/stable/user/basics.subclassing.html#the-role-of-array-finalize
+
+        Parameters
+        ----------
+        obj : MPIArray or None
+            The original MPIArray being viewed or broadcast. When in the middle of a constructor, obj is set to None.
+        """
+        # we are in the middle of a constructor, and the attributes
+        # will be set when we return to it
+        if obj is None:
+            return
+
+        # we are in a ufunc, rebuild the attributes from the original MPIArray
+        if isinstance(obj, MPIArray):
+            comm = getattr(obj, "comm", mpiutil.world)
+            axis = getattr(
+                obj, "axis", 0
+            )  # probably not a good default! How would we find this out?
```

If you can't reconstruct the distributed axis then it should throw an exception rather than guess.